src/python_testing/TC_IDM_4_2.py

```diff
@@ -0,0 +1,587 @@
+        asserts.assert_true(sub_cr1_step2_max_interval_ceiling_sec <= max_interval_ceiling_sec,
+                            "MaxInterval is not less than or equal to MaxIntervalCeiling")
+
+        sub_cr1_step2.Shutdown()
+
+        '''
+        ##########
+        Step 3
+        ##########
+        '''
+        self.print_step(3, "Setup CR2 such that it does not have access to a specific cluster. CR2 sends a subscription message to subscribe to an attribute on that cluster for which it does not have access.")
+
+        # Limited ACE for controller 2 with single cluster access
+        CR2_limited_ace = Clusters.AccessControl.Structs.AccessControlEntryStruct(
+            privilege=Clusters.AccessControl.Enums.AccessControlEntryPrivilegeEnum.kView,
+            authMode=Clusters.AccessControl.Enums.AccessControlEntryAuthModeEnum.kCase,
+            targets=[Clusters.AccessControl.Structs.AccessControlTargetStruct(cluster=Clusters.BasicInformation.id)],
+            subjects=[CR2_nodeid])
+
+        # Add controller 2 limited ACE to DUT ACL
+        dut_acl = copy.deepcopy(dut_acl_original)
+        dut_acl.append(CR2_limited_ace)
+
+        # Write updated DUT ACL into DUT
+        await CR1.WriteAttribute(
+            nodeid=self.dut_node_id,
+            attributes=[(0, Clusters.AccessControl.Attributes.Acl(dut_acl))]
+        )
+
+        # Controller 2 tries to subscribe an attribute from a cluster
+        # it doesn't have access to
+        # "INVALID_ACTION" status response expected
+        try:
+            await CR2.ReadAttribute(
+                nodeid=self.dut_node_id,
+                # Attribute from a cluster controller 2 has no access to
+                attributes=[(0, Clusters.AccessControl.Attributes.Acl)],
+                keepSubscriptions=False,
+                reportInterval=[3, 3],
+                autoResubscribe=False
+            )
+        except ChipStackError as e:
+            # Verify that the DUT returns an "INVALID_ACTION" status response
+            asserts.assert_equal(e.err, INVALID_ACTION_ERROR_CODE,
+                                 "Incorrect error response for subscription to unallowed cluster")
+
+        '''
+        ##########
+        Step 4
+        ##########
+        '''
+        self.print_step(4, "Setup CR2 such that it does not have access to all attributes on a specific cluster and endpoint. CR2 sends a subscription request to subscribe to all attributes for which it does not have access.")
+
+        # Limited ACE for controller 2 with single cluster access
+        CR2_limited_ace = Clusters.AccessControl.Structs.AccessControlEntryStruct(
+            privilege=Clusters.AccessControl.Enums.AccessControlEntryPrivilegeEnum.kView,
+            authMode=Clusters.AccessControl.Enums.AccessControlEntryAuthModeEnum.kCase,
+            targets=[Clusters.AccessControl.Structs.AccessControlTargetStruct(cluster=Clusters.BasicInformation.id)],
```

Comment:
This is a poorly written test case, but I THINK the intent here is to ensure that you can't subscribe to a cluster if the endpoint is different in the ACL. So can you add endpoint=1 here, then try to read from the basic information cluster on EP0?